### PR TITLE
migration state: add source pod and failure reason

### PR DESCRIPTION
**Pull request by jean-edouard**
### What this PR does
Before this PR:
Migration state has no source pod and no failure reason

After this PR:
Migration state has a source pod and a failure reason

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
More information in the migration state of VMI / migration objects
```

### CODE REVIEW

**Comment by jean-edouard** [comment]
/hold
I broke unit tests

File: pkg/virt-controller/watch/migration.go [hunk 1]
```diff
@@ -837,7 +837,8 @@ func (c *MigrationController) handleMarkMigrationFailedOnVMI(migration *virtv1.V
 		return err
 	}
 	log.Log.Object(vmi).Infof("Marked Migration %s/%s failed on vmi due to target pod disappearing before migration kicked off.", migration.Namespace, migration.Name)
-	c.recorder.Event(vmi, k8sv1.EventTypeWarning, FailedMigrationReason, fmt.Sprintf("VirtualMachineInstance migration uid %s failed. reason: target pod is down", string(migration.UID)))
+	vmiCopy.Status.MigrationState.FailureReason = "Target pod is down"
+	c.recorder.Event(vmi, k8sv1.EventTypeWarning, FailedMigrationReason, fmt.Sprintf("VirtualMachineInstance migration uid %s failed. reason: %s", string(migration.UID), vmiCopy.Status.MigrationState.FailureReason))
```
Comment (vladikr) [inline pkg/virt-controller/watch/migration.go:841]: Is there a way that we will already have a Failure Reason from virt-handler at this point?

Comment (jean-edouard) [reply pkg/virt-controller/watch/migration.go:841]: Yeah that's possible. What do you suggest?

Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go:841]: Wouldn't creating the event for the migration object solve this?

File: tests/migration/migration.go [hunk 2]
```diff
@@ -316,10 +316,12 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 
 			// execute a migration, wait for finalized state
 			By("Starting the Migration for iteration")
+			sourcePod := tests.GetRunningPodByVirtualMachineInstance(vmi, testsuite.GetTestNamespace(vmi))
 			migration := libmigration.New(vmi.Name, vmi.Namespace)
 			migration = libmigration.RunMigrationAndExpectToCompleteWithDefaultTimeout(virtClient, migration)
 			By("Checking VMI, confirm migration state")
-			libmigration.ConfirmVMIPostMigration(virtClient, vmi, migration)
+			vmi = libmigration.ConfirmVMIPostMigration(virtClient, vmi, migration)
+			Expect(vmi.Status.MigrationState.SourcePod).To(Equal(sourcePod.Name))
```
Comment (Barakmor1) [inline tests/migration/migration.go:324]: ExpectWithOffset would be better in that case

File: pkg/virt-controller/watch/migration.go (hunk 1 above)
Comment (vladikr) [reply pkg/virt-controller/watch/migration.go:841]: @Barakmor1 I think these are two separate records. We need to store this info in the status for convenience. That way everything related to this migration is stored in one place.

Comment (vladikr) [reply pkg/virt-controller/watch/migration.go:841]: > Yeah that's possible. What do you suggest?

@jean-edouard I think we can merge. To simply check whether `MigrationState.FailureReason` has something already and add this message. wdyt?

File: tests/migration/migration.go (hunk 2 above)
Comment (jean-edouard) [reply tests/migration/migration.go:324]: The rest of the function doesn't use `WithOffset`, and I would argue that it's a good thing.
`WithOffset` effectively obfuscates which line of the function tripped. and this is a massive function called once by 2 different tests.
The code for the first test that uses the function is 2 lines long, and the second one has just 1 line: the call to this function!
So, in this case, moving the failure up one level effectively makes the line number meaningless.
WDYT?

Comment (Barakmor1) [reply tests/migration/migration.go:324]: Sounds good :)

**Comment by vladikr** [comment]
/approve

Looks good to me.

**Comment by vladikr** [comment]
Thanks @jean-edouard !

**Comment by vladikr** [comment]
@jean-edouard should we back port this?

**Pull request merged**